### ISSUE TITLE
spam

### ISSUE THREAD

**Issue by muling** [wet-boew/wet-boew#8729]

**Comment by thekodester** [comment]
WET doesn't have a spam plugin, but I would recommend https://www.google.com/recaptcha/intro/v3.html it pretty much what is used across the web.

**Comment by thekodester** [comment]
You can set the difficulty of the image matching and for non visual users it has a built-in audio option.

**Comment by thekodester** [comment]
The only cons is it requires a google account. So your team would need to internally keep track of the email and password used for it.

**Comment by muling** [comment]
We definitely want to not rely on Google as we are trying to move away from their services.

**Comment by muling** [comment]
As well I was hoping for a solution that isn't reliant on JS.

**Comment by thekodester** [comment]
You could go old school. Have a hidden field. Name it something like "Website". If the field has data reject the form. It not the best method, but it one of the few not reliant on JS.

**Comment by thekodester** [comment]
Reason why I say name it like "Website" as bots are more like to fill it out as the want you to click on a website or url/link.

**Comment by thekodester** [comment]
Needs to be style="display:none".

**Comment by muling** [comment]
Yes we've tried the honey pot method. We are still experimenting with it.  
With display:none, the bot won't see it, so it has to be visible but hidden from the screen reader.

**Comment by thekodester** [comment]
The other method is having a spam keyword search. You check the data submitted for things like "Free investment" or "100% free" or "Selling" and so on.

**Comment by thekodester** [comment]
If it matches reject the submission, but you can run into false positives.

**Comment by muling** [comment]
I don't think I want to rely on blacklisting.  That is like going down a rabbit hole.

**Comment by thekodester** [comment]
If you have tried "honey pot" (which well work with display:none for some bots) or "spam keyword/blacklisting" methods then you have to use a javascript method. There a many free alternatives to recaptcha.

Otherwise you just have to take down the form.

**Comment by muling** [comment]
so no one can contact us...lol.

**Comment by muling** [comment]
I'm hoping adding a 'timestamp' solution on top of honey pot would help make it more effective.

**Comment by muling** [comment]
So Canada.ca doesn't use any spam filtering?

**Comment by muling** [comment]
i think the bots can do simple math problems. it's not hard to parse that out and do the calculation.

**Comment by thekodester** [comment]
You can make it more complicated. It just an example.

**Comment by thekodester** [comment]
Would be very hard to code depending on the server side language.

**Comment by muling** [comment]
PHP.

**Comment by thekodester** [comment]
PHP would be very easy either using cookies, but best with session variables.

**Comment by muling** [comment]
Thanks for your suggestions.

**Comment by thekodester** [comment]
Found this online: https://carlofontanos.com/adding-a-simple-math-captcha-to-your-form-in-php/

**Comment by thekodester** [comment]
Some alterations would be need to make it more accessible.

**Comment by muling** [comment]
Do you know who is in charge of the Canada.ca site?

**Comment by thekodester** [comment]
No, but the main people here are @EricDunsworth, @GormFrank, and @duboisp. Believe they are still employed by the government.

**Comment by muling** [comment]
Thanks.

**Comment by muling** [comment]
It would be nice to hear if they have any suggestions.

**Comment by thekodester** [comment]
One of them we'll comment if they do. They have been tagged.

**Comment by EricDunsworth** [comment]
Maybe something along these lines could do the trick? Although it'd depend on how unsophisticated the spam bots you're dealing with are.

**Code sample:**
```html
<div class="wb-inv" aria-hidden>
	<label for "url123">Ignore this field and leave it empty</label>
	<input type="url" type="url123" name="url123" tabindex="-1" />
</div>
```

**How it works:**
* ``class="wb-inv"`` visually-hides the ``<div>``'s contents in WET (doesn't rely on ``display: none;`` or ``visibility: hidden;``)
  * ``class="sr-only"`` should also work in both WET and Bootstrap
* ``aria-hidden`` tells assistive technologies (ATs) to ignore the ``<div>``'s contents
* ``type="url"`` input field baits spam bots into filling in the URL they're likely trying to spread
* ``tabindex="-1"`` prevents the ``<input>`` field from being focusable
* The label's name directs users to leave the field blank in the highly unlikely event of an AT running into it

**Sources:**
* My own knowledge/quick testing in a few browsers
* [Stack Overflow: Aria-hidden on a Honey Pot field](https://stackoverflow.com/questions/34026253/aria-hidden-on-a-honey-pot-field)
* [Ask HN: A Good Alternative for ReCaptcha?](https://news.ycombinator.com/item?id=20058697)

**Comment by muling** [comment]
Yes we are already using wb-inv, but we wanted to add the aria-hidden and test that out.  Our label is similar to what you have there.

**Comment by duboisp** [comment]
> Do you know who is in charge of the Canada.ca site

It's my team and I am the dev lead.

> So Canada.ca doesn't use any spam filtering?
Canada.ca is in the cloud and our CDN provider do a lot for us to prevent spam.

On the CAPTCHA note,
It would be possible to have a CAPTCHA in wet-boew as long it is conform to WCAG 2.1 Level AA standard. In the [SC 1.1.1 Understanding](https://www.w3.org/WAI/WCAG21/Understanding/non-text-content.html) there is a note about CAPTCHA and if a such feature is developed as a WET-BOEW plugin by the Principal Publisher or by the community, I am expecting we will go beyond the minimal requirement as it is recommended.

If you would like to discuss in person or want some guidance on how you can develop a wet-boew CAPTCHA plugin, you can come to meet us at one of the weekly WET-BOEW code sprint. It's on every Tuesday from 9h30 to 3h30 at 200 du portage in Gatineau room 0990.

**Comment by thekodester** [comment]
Also on this topic. Server side you can make sure your following the https://en.wikipedia.org/wiki/Post/Redirect/Get method so a bot cannot just continually refresh the page and resubmit the form.

**Comment by muling** [comment]
Thanks for the info.

**Comment by muling** [comment]
I'm not in the NCR.
